app/extensions/__init__.py

```diff
@@ -79,6 +80,50 @@
 ##########################################################################################
 
 
+class ExtraValidationSchema(Schema):
+    @validates_schema(pass_original=True)
+    def validates_schema(self, data, original_data):
+        """
+        Raise validation error if there are extra fields not defined in
+        the schema.
+
+        This is necessary because marshmallow (before 3.13.0) just
+        ignores extra fields without any validation errors.
+        """
+        if data is None:  # something wrong with the data, skip
+            return
+        valid_fields = sorted(self.fields.keys())
+        unknown = set(original_data) - set(valid_fields)
+        if unknown:
+            raise ValidationError(
+                f'Unknown field(s): {", ".join(unknown)}, options are {", ".join(valid_fields)}'
+            )
+
+    def get_error_message(self, errors):
+        error_keys = list(errors.keys())
+        if error_keys == ['_schema']:
+            return ' '.join(errors['_schema'])
+
+        def get_error(errors, results, _keys=[]):
```

Comment:
I can see that this is called from within the model code when it detects that there is an error but can you add some description about what this is doing. I think I get what this is doing in that the validates_schema is called fro the top level and all 'child' schemas but there's far too much guesswork, can you add some more description please. 
Especially when code has recursion, I think it's important to document why it's needed.

Reply:
Validation errors are like this:
         
```
{
    'all': {
        '_schema': [
            'Unknown field(s): random, options are email, restAPI'
        ],
        'restAPI': [
            'Not a valid boolean.'
        ]
    }
}
```
                                                          
This method turns this into an error message like:
                                                                    
```
"all": Unknown field(s): random, options are email, restAPI. "all.restAPI": Not a valid boolean.
```

^ I'll add something like this in the comments

Reply:
```python
        def get_error(errors, results, _keys=[]):
            # Traverse down the error structure depth first so we can get the
            # actual field name.  For example in the example above, we want
            # "all.restAPI" as the field name.
```